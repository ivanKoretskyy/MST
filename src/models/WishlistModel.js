import { types } from "mobx-state-tree";

export const WishListItemModel = types.model({
  name: types.string,
  price: types.number,
  image: ''
}).actions(self => ({
  changeName: (newName) => {
    self.name = newName;
  },
  changePrice: (newPrice) => {
    self.price = newPrice;
  },
  changeImage: (newImage) => {
    self.image = newImage
  }
}));

export const WishListModel = types.model({
  list: types.optional(types.array(WishListItemModel), [])
})
.actions(self => ({
  addWishlist: (wishlist) => {
    self.list.push(wishlist)
  }
}))
.views(self => ({
  get totalPrice() {
    return self.list.reduce((acc, element) => acc + element.price, 0)
  }
}))