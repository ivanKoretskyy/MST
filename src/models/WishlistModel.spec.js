import { WishListItemModel, WishListModel } from './WishlistModel';

import { getSnapshot, onSnapshot, onPatch } from 'mobx-state-tree';

describe('modle', ()=> {
  it('should create item',() => {
    const item = WishListItemModel.create({
      name: 'car',
      price: 34
    })
    expect(item.price).toEqual(34)
  })

  it('should change name', () => {
    const item = WishListItemModel.create({
      name: 'car',
      price: 34
    });

    item.changeName('House');
    expect(item.name).toEqual('House');
  })

  it('should change price', () => {
    const item = WishListItemModel.create({
      name: 'car',
      price: 34
    });

    item.changePrice(66);
    expect(item.price).toEqual(66);
  })

// --------------------------------list tests----------------------
  it('should create list',() => {
    const items = WishListModel.create({
      list: [{      name: 'car',
      price: 34}]
    })
    expect(items.list[0].price).toEqual(34)
  })


  it('should add item to  list',() => {
    const items = WishListModel.create()
    const states = []
    onSnapshot(items, snapshot => {
      states.push(snapshot);
    })

    const patches = []
    onPatch(items, patch => {
      patches.push(patch);
    })

    items.addWishlist(WishListItemModel.create({
      name: 'car',
      price: 34
    }));

    expect(items.list[0].price).toEqual(34)
    items.list[0].changePrice(66);
    expect(items.list[0].price).toEqual(66)
    expect(getSnapshot(items)).toMatchSnapshot();
    expect(states).toMatchSnapshot();
    expect(patches).toMatchSnapshot();
  })
})