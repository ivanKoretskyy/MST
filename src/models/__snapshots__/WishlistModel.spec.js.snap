// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modle should add item to  list 1`] = `
Object {
  "list": Array [
    Object {
      "image": "",
      "name": "car",
      "price": 66,
    },
  ],
}
`;

exports[`modle should add item to  list 2`] = `
Array [
  Object {
    "list": Array [
      Object {
        "image": "",
        "name": "car",
        "price": 34,
      },
    ],
  },
  Object {
    "list": Array [
      Object {
        "image": "",
        "name": "car",
        "price": 66,
      },
    ],
  },
]
`;

exports[`modle should add item to  list 3`] = `
Array [
  Object {
    "op": "add",
    "path": "/list/0",
    "value": Object {
      "image": "",
      "name": "car",
      "price": 34,
    },
  },
  Object {
    "op": "replace",
    "path": "/list/0/price",
    "value": 66,
  },
]
`;
